consensus/hotstuff/timeoutcollector/timeout_processor.go

```diff
@@ -0,0 +1,295 @@
+			return model.NewInvalidTimeoutErrorf(timeout, "timeout must include TC")
+		}
+	}
+
+	// 2. Check fi signer identity is valid
+	_, err := p.committee.IdentityByEpoch(timeout.View, timeout.SignerID)
+	if model.IsInvalidSignerError(err) {
+		return model.NewInvalidTimeoutErrorf(timeout, "invalid signer for timeout: %w", err)
+	}
+	if err != nil {
+		return fmt.Errorf("error retrieving signer Identity at view %d: %w", timeout.View, err)
+	}
+
+	// 3. Check if QC is valid
+	err = p.validator.ValidateQC(timeout.NewestQC)
+	if err != nil {
+		if model.IsInvalidQCError(err) {
+			return model.NewInvalidTimeoutErrorf(timeout, "included QC is invalid: %w", err)
+		}
+		if errors.Is(err, model.ErrViewForUnknownEpoch) {
+			// We require each replica to be bootstrapped with a QC pointing to a finalized block. Therefore, we should know the
+			// Epoch for any QC.View and TC.View we encounter. Receiving a `model.ErrViewForUnknownEpoch` is conceptually impossible,
+			// i.e. a symptom of an internal bug or invalid bootstrapping information.
+			return fmt.Errorf("no Epoch information availalbe for QC that was included in TO; symptom of internal bug or invalid bootstrapping information: %s", err.Error())
+		}
+		return fmt.Errorf("unexpected error when validating QC: %w", err)
+	}
+
+	// 4. If TC is included, it must be valid
+	if timeout.LastViewTC != nil {
+		err = p.validator.ValidateTC(timeout.LastViewTC)
+		if err != nil {
+			if model.IsInvalidTCError(err) {
+				return model.NewInvalidTimeoutErrorf(timeout, "included TC is invalid: %w", err)
+			}
+			if errors.Is(err, model.ErrViewForUnknownEpoch) {
+				// We require each replica to be bootstrapped with a QC pointing to a finalized block. Therefore, we should know the
+				// Epoch for any QC.View and TC.View we encounter. Receiving a `model.ErrViewForUnknownEpoch` is conceptually impossible,
+				// i.e. a symptom of an internal bug or invalid bootstrapping information.
+				return fmt.Errorf("no Epoch information availalbe for TC that was included in TO; symptom of internal bug or invalid bootstrapping information: %s", err.Error())
+			}
+			return fmt.Errorf("unexpected error when validating TC: %w", err)
+		}
+	}
+	return nil
+
+}
+
+// buildTC performs aggregation of signatures when we have collected enough
+// weight for building TC. This function is run only once by single worker.
+// Any error should be treated as exception.
+func (p *TimeoutProcessor) buildTC() (*flow.TimeoutCertificate, error) {
+	signers, highQCViews, aggregatedSig, err := p.sigAggregator.Aggregate()
+	if err != nil {
+		return nil, fmt.Errorf("could not aggregate multi message signature: %w", err)
+	}
+
+	signerIndices, err := p.signerIndicesFromIdentities(signers)
+	if err != nil {
+		return nil, fmt.Errorf("could not encode signer indices: %w", err)
+	}
+
+	return &flow.TimeoutCertificate{
+		View:          p.view,
+		NewestQCViews: highQCViews,
+		NewestQC:      p.NewestQCTracker.NewestQC(),
```

Comment:
## ❗ potential race condition ❓ 

I am not entirely sure, but to me this looks like a potential edge case, where a race condition could lead to constructing an invalid TC. Reasoning:
* In the `Validator`, we require that the QC included in a TC satisfies `QC.View` must be equal to `max(TC.NewestQCViews)`:  https://github.com/onflow/flow-go/blob/90b4f79a88d5bcfdbfa11f07d80e7144df59cf40/consensus/hotstuff/validator/validator.go#L88-L97
* Specifically, this means that it is _illegal_ to include a newer QC or an older QC
* I think this code is vulnerable to a race condition: https://github.com/onflow/flow-go/blob/90b4f79a88d5bcfdbfa11f07d80e7144df59cf40/consensus/hotstuff/timeoutcollector/timeout_processor.go#L137-L151
   * I managed to sketch an argument in my head that including a newer QC is impossible. This is because the newest QC known to the `NewestQCTracker` must previously have been added to the `sigAggregator`: https://github.com/onflow/flow-go/blob/90b4f79a88d5bcfdbfa11f07d80e7144df59cf40/consensus/hotstuff/timeoutcollector/timeout_processor.go#L146-L151 
   * However, I don't think the reverse holds. Lets consider the following situation:
      * There is only one honest node (lets call that node 'Alice') that knows the highest QC for view 100. All other nodes only know the lower QC for view 99.
      * We receive TOs from enough nodes to construct a TC, without Alice's TO. 
      * `Thread 1` is in [line 145](https://github.com/onflow/flow-go/blob/90b4f79a88d5bcfdbfa11f07d80e7144df59cf40/consensus/hotstuff/timeoutcollector/timeout_processor.go#L145) with the _last_ TO (_not_ Alice's TO) needed to construct a TC.
      * `Thread 2` also sits in [line 145](https://github.com/onflow/flow-go/blob/90b4f79a88d5bcfdbfa11f07d80e7144df59cf40/consensus/hotstuff/timeoutcollector/timeout_processor.go#L145) with Alice's TO.
      * `Thread 1` gets to go first. It adds the last needed TO to `sigAggregator` and the returned `totalWeight` meets the minimal threshold for the first time. Therefore, `Thread 1` passes [line 160](https://github.com/onflow/flow-go/blob/90b4f79a88d5bcfdbfa11f07d80e7144df59cf40/consensus/hotstuff/timeoutcollector/timeout_processor.go#L160. It reaches [line 163](https://github.com/onflow/flow-go/blob/90b4f79a88d5bcfdbfa11f07d80e7144df59cf40/consensus/hotstuff/timeoutcollector/timeout_processor.go#L163) and pauses there. 
      * Now, `Thread 2` executes [line 146](https://github.com/onflow/flow-go/blob/90b4f79a88d5bcfdbfa11f07d80e7144df59cf40/consensus/hotstuff/timeoutcollector/timeout_processor.go#L146), adding Alice's TO to `sigAggregator`. But `Thread 2` pauses in [line 150](https://github.com/onflow/flow-go/blob/90b4f79a88d5bcfdbfa11f07d80e7144df59cf40/consensus/hotstuff/timeoutcollector/timeout_processor.go#L150).
      * At this point, `sigAggregator` knows that there is a TO for view 100 (Alice's TO). But the TO has not been added to the `NewestQCTracker`.
      * Finally, `Thread 1` enters [`p.buildTC()`](https://github.com/onflow/flow-go/blob/90b4f79a88d5bcfdbfa11f07d80e7144df59cf40/consensus/hotstuff/timeoutcollector/timeout_processor.go#L164), aggregates the signatures and includes view 100 in the `TC.NewestQCViews`. However, `p.NewestQCTracker.NewestQC()` only knows a QC for view 99, which we include in the TC. 

Hence, we have constructed an invalid TC ☠️

Reply:
From my perspective, the root cause for this edge case is that tracking the views and the highest QC is implemented as a non-atomic operation: https://github.com/onflow/flow-go/blob/90b4f79a88d5bcfdbfa11f07d80e7144df59cf40/consensus/hotstuff/timeoutcollector/timeout_processor.go#L146-L151

I can think of two different ways around this problem:
* In both proposals, we would switch the order of operations:
    ```golang
    p.NewestQCTracker.Track(timeout.NewestQC)

    totalWeight, err := p.sigAggregator.VerifyAndAdd(timeout.SignerID, timeout.SigData, timeout.NewestQC.View)
    if err != nil {
      return fmt.Errorf("could not process invalid signature: %w", err)
    }
    ```
   Then, `NewestQCTracker` could know of a newer QC than `sigAggregator`
* _Proposal 1_:  
   Based on my understanding, the HotStuff proofs require that a TC includes a QC that has view _at least as new_ as `max(TC.NewestQCViews)`. However, I think it _might_ be possible to allow a **newer** QC, with view strictly higher than any value in `TC.NewestQCViews`. Though, this is only a hypothesis; we would need to go over the proofs in detail to confirm.
* _Proposal 2_:  
   The other option would be that `NewestQCTracker` does not only track the QC with largest view, but stores _all_ QCs that were included in TOs for the respective view. Well, technically we only need one QC for each view. For example, `NewestQCTracker` could maintain a map: `map[uint64]QuorumCertificate`. In `buildTC()`, we would then retrieve the specific QC that matches `max(highQCViews)`.

Reply:
Thanks for spotting it, really good observation, very nice. I have a feeling proposal 1 is perfectly fine, for safety properties in all proofs we build proofs based on values of `TC.NewestQCViews`. we will vote for proposal if `QC.View >= max(TC.NewestQCViews)`. We already allow including newer QC into proposal. 
Let's consider a case where TC includes QC: `TC.NewestQC.View > max(Tc.NewestQCViews)`. 
1. Leader includes includes that QC into proposal, this satisfies our rule `QC.View >= max(TC.NewestQCViews)` -> it's safe to vote for this proposal.
2. Leader doesn't include that QC into proposal, then `QC.View >= max(TC.NewestQCViews)` holds, but this proposal is not safe to vote for since we fail this [check](https://github.com/onflow/flow-go/blob/feature/active-pacemaker/consensus/hotstuff/safetyrules/safety_rules.go#L257)

Based on proofs we require only that `QC.View >= max(TC.NewestQCViews)` but in our implementation we need to make sure that we are extending the newest QC available.

Reply:
Another option: we could embed the functionality of `NewestQCTracker` within `TimeoutSignatureAggregator` and update both `NewestQC` and `NewestQCViews` atomically in the locked section here:
https://github.com/onflow/flow-go/blob/338166875960fde9396caeb670445a6f16df6eb2/consensus/hotstuff/timeoutcollector/aggregation.go#L126-L131

Reply:
@jordanschalm I am pretty sure we have decided to keep QC tracking and sig aggregation as separate paths, maybe we need to revisit this

Reply:
Result of discussion: 
* we'll implement Proposal 1 (allowing `QC.View >= max(TC.NewestQCViews)` as opposed to requiring strict equality)